hugr-passes/src/monomorphize.rs

```diff
@@ -0,0 +1,468 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use hugr_core::{
+    extension::ExtensionRegistry,
+    ops::{Call, FuncDefn, OpTrait},
+    types::{Signature, Substitution, TypeArg},
+    Node,
+};
+
+use hugr_core::hugr::{hugrmut::HugrMut, internal::HugrMutInternals, Hugr, HugrView, OpType};
+
+/// Replaces calls to polymorphic functions with calls to new monomorphic
+/// instantiations of the polymorphic ones.
+///
+/// If the Hugr is [Module](OpType::Module)-rooted,
+/// * then the original polymorphic [FuncDefn]s are left untouched (including Calls inside them)
+///     - call [remove_polyfuncs] when no other Hugr will be linked in that might instantiate these
+/// * else, the originals are removed (they are invisible from outside the Hugr).
+pub fn monomorphize(mut h: Hugr, reg: &ExtensionRegistry) -> Hugr {
+    let root = h.root();
+    // If the root is a polymorphic function, then there are no external calls, so nothing to do
+    if !is_polymorphic_funcdefn(h.get_optype(root)) {
+        mono_scan(&mut h, root, None, &mut HashMap::new(), reg);
+        if !h.get_optype(root).is_module() {
+            return remove_polyfuncs(h);
+        }
+    }
+    h
+}
+
+/// Removes any polymorphic [FuncDefn]s from the Hugr. Note that if these have
+/// calls from *monomorphic* code, this will make the Hugr invalid (call [monomorphize]
+/// first).
+pub fn remove_polyfuncs(mut h: Hugr) -> Hugr {
```

Comment:
Hmmm, I like your thinking here - there is nothing special about *polymorphic* functions that means they should be removed, just that they have no uses.

However, this requires doing a reachability/call-graph analysis, the criteria you suggest are not sufficient to remove mutually-recursive functions. (There are also similar issues to those we have discussed before about library entrypoints, `public: bool` etc.)

How about we punt the whole lot to #1672 and don't do any FuncDefn-removal at all in this PR?

Reply:
Or a simpler, static/call-graph notion of reachability, rather than the dynamic one of #1672 (so, assuming every conditional can branch both ways, etc.) could be more appropriate - if we do the more complex dataflow analysis, it's a lot more work to patch up the Hugr rather than just deleting some functions.

However, still needs reachability (ideally, would allow arbitrary functions to be designated reachable in a library, *including polymorphic ones*). Whereas here we "piggyback" off reachability done by monomorphization....